apps/emqx/src/emqx_channel.erl

```diff
@@ -3192,6 +3210,22 @@ run_hooks(Name, Args, Acc) ->
     ok = emqx_metrics:inc(Name),
     emqx_hooks:run_fold(Name, Args, Acc).
 
+-compile({inline, [run_hook_with_context/3]}).
+run_hook_with_context(Name, Args, Channel) ->
+    HookContext = mk_common_hook_context(Channel),
```

Comment:
we will need some perf test to check if this will work well.
the session object for example is non-trivial term to copy into pd for each message.